lib/protobuf/generators/file_generator.rb

```diff
@@ -142,6 +179,75 @@ def fully_qualified_token?(token)
         token[0] == '.'
       end
 
+      def eval_dependencies(name, namespace = nil)
+        name = "#{namespace}.#{name}" if namespace && !fully_qualified_token?(name)
+        return if name.empty? || @evaled_dependencies.include?(name) || modulize(name).safe_constantize
+
+        # if name = .foo.bar.Baz look for classes / modules named ::Foo::Bar and ::Foo
+        # module == pure namespace (e.g. the descriptor package name)
+        # class == nested messages
+        create_ruby_namespace_heiarchy(name)
+
+        if (message = @all_messages[name])
+          # Create the blank namespace in case there are nested types
+          eval_message_code(name)
+
+          message.nested_type.each do |nested_type|
+            eval_dependencies(nested_type.name, name) unless nested_type.name.empty?
+          end
+          message.field.each do |field|
+            eval_dependencies(field.type_name, name) unless field.type_name.empty?
+          end
+          message.enum_type.each do |enum_type|
+            eval_dependencies(enum_type.name, name)
+          end
+
+          # Check @evaled_dependencies again in case there was a dependency
+          # loop that already loaded this message
+          return if @evaled_dependencies.include?(name)
+          eval_message_code(name, message.field)
+          @evaled_dependencies << name
+
+        elsif (enum = @all_enums[name])
+          # Check @evaled_dependencies again in case there was a dependency
+          # loop that already loaded this enum
+          return if @evaled_dependencies.include?(name)
+          namespace = name.split(".")
+          eval_enum_code(enum, namespace[0..-2].join("."))
+          @evaled_dependencies << name
+        else
+          raise "Error loading unknown dependencies, could not find message or enum #{name.inspect}"
+        end
+      end
+
+      def eval_message_code(fully_qualified_namespace, fields = [])
+        group = GroupGenerator.new(0)
+        group.add_extended_messages({ fully_qualified_namespace => fields }, false)
+        print group.to_s
+        eval_code
+      end
+
+      def eval_enum_code(enum, fully_qualified_namespace)
+        group = GroupGenerator.new(0)
+        group.add_enums([enum], :namespace => [fully_qualified_namespace])
+        print group.to_s
+        eval_code(modulize(fully_qualified_namespace).safe_constantize || Object)
+      end
+
+      def eval_code(context = Object)
+        warn "#{context.inspect}.module_eval #{print_contents.inspect}" if ENV['PB_DEBUG']
+        context.module_eval print_contents.to_s
+        @io.truncate(0)
+        @io.rewind
+      end
+
+      def create_ruby_namespace_heiarchy(name)
```

Comment:
The code in this method took me a while to understand. I think you can use `rpartition` to achieve the same results:

```
def create_ruby_namespace_heiarchy(namespace)
  loop do
    namespace, _match, _tail = namespace.rpartition(".")
    break if namespace.empty?
    puts namespace
  end
end

irb(main):142:0> create_ruby_namespace_heiarchy('test.pkg.code_generator_spec')
test.pkg
test
=> nil
irb(main):143:0> create_ruby_namespace_heiarchy(".google.protobuf.FieldOptions")
.google.protobuf
.google
```